### PR TITLE
Adding 'Raises' sections to docstrings throughout PL

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**
The ``Raises`` section is often missing from docstrings in PennyLane. Docstrings modifiend in the upcoming PR  #377 were left intact.

**Description of the Change:**
Adds the ``Raises`` section with a related message.

**Benefits:**
Information about exceptions appears in the documentation online.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: pennylane/_device.py [hunk 1]
```diff
@@ -116,6 +116,9 @@ def shots(self, shots):
         Args:
             shots (int): number of circuit evaluations/random samples used to estimate
                 expectation values of observables
+
+        Raises:
+            DeviceError: if number of `shots` is less than 1
```
Comment (josh146) [inline pennylane/_device.py]:
```suggestion
            DeviceError: if number of shots is less than 1
```

File: pennylane/_device.py [hunk 2]
```diff
@@ -146,6 +149,9 @@ def execute(self, queue, observables, parameters={}):
             parameters (dict[int->list[ParameterDependency]]): Mapping from free parameter index to the list of
                 :class:`Operations <pennylane.operation.Operation>` (in the queue) that depend on it.
 
+        Raises:
+            QuantumFunctionError: if the return type of one of the `observables` is not supported
```
Comment (josh146) [inline pennylane/_device.py]:
```suggestion
            QuantumFunctionError: if the value of :attr:`~.Observable.return_type` is not supported
```

File: pennylane/qnode.py [hunk 3]
```diff
@@ -186,6 +190,11 @@ def construct(self, args, kwargs=None):
                 however PennyLane does not support differentiating with respect to keyword arguments.
                 Instead, keyword arguments are useful for providing data or 'placeholders'
                 to the quantum circuit function.
+
+        Raises:
+            QuantumFunctionError: if the :class:`pennylane.QNode`'s _current_context is attempted to be modified
+                inside of this method, the quantum function returns incorrect values or if
+                both continuous and discrete operations are specified in the same quantum circuit
```
Comment (josh146) [inline pennylane/qnode.py]: Note that changes here will also need to be duplicated for the new QNode once merged :)

File: pennylane/qnode.py [hunk 4]
```diff
@@ -767,6 +782,11 @@ def jacobian(self, params, which=None, *, method='B', h=1e-7, order=1, **kwargs)
             shots (int): How many times the circuit should be evaluated (or sampled) to estimate
                 the expectation values.
 
+        Raises:
+            QuantumFunctionError: if sampling is specified
+            ValueError: if `params` is incorrect, the differentiation is not possible or if `method`
```
Comment (josh146) [inline pennylane/qnode.py]:
```suggestion
            ValueError: if `params` is incorrect, the differentiation is not possible, or if `method`
```

File: pennylane/templates/embeddings.py [hunk 5]
```diff
@@ -48,6 +48,9 @@ def AngleEmbedding(features, wires, rotation='X'):
 
     Keyword Args:
         rotation (str): Type of rotations used
+
+    Raises:
+        ValueError: if `features` or `wires` is invalid
```
Comment (josh146) [inline pennylane/templates/embeddings.py]:
```suggestion
        ValueError: if ``features`` or ``wires`` is invalid
```

File: pennylane/templates/embeddings.py [hunk 6]
```diff
@@ -92,6 +95,9 @@ def AmplitudeEmbedding(features, wires, pad=False, normalize=False):
         wires (Sequence[int]): sequence of qubit indices that the template acts on
         pad (Boolean): controls the activation of the padding option
         normalize (Boolean): controls the activation of automatic normalization
+
+    Raises:
+        ValueError: if `features` or `wires` is invalid
```
Comment (josh146) [inline pennylane/templates/embeddings.py]:
```suggestion
        ValueError: if ``features`` or ``wires`` is invalid
```

File: pennylane/templates/embeddings.py [hunk 7]
```diff
@@ -146,6 +152,9 @@ def BasisEmbedding(features, wires):
     Args:
         features (array): Binary input array of shape ``(n, )``
         wires (Sequence[int]): sequence of qubit indices that the template acts on
+
+    Raises:
+        ValueError: if `features` or `wires` is invalid
```
Comment (josh146) [inline pennylane/templates/embeddings.py]:
```suggestion
        ValueError: if ``features`` or ``wires`` is invalid
```

File: pennylane/templates/embeddings.py [hunk 8]
```diff
@@ -181,6 +190,9 @@ def SqueezingEmbedding(features, wires, method='amplitude', c=0.1):
             ``'amplitude'`` uses the amplitude
         c (float): value of the phase of all squeezing gates if ``execution='amplitude'``, or the
             amplitude of all squeezing gates if ``execution='phase'``
+
+    Raises:
+        ValueError: if `features` or `wires` is invalid
```
Comment (josh146) [inline pennylane/templates/embeddings.py]:
```suggestion
        ValueError: if ``features`` or ``wires`` is invalid
```

File: pennylane/templates/embeddings.py [hunk 9]
```diff
@@ -222,6 +234,9 @@ def DisplacementEmbedding(features, wires, method='amplitude', c=0.1):
             ``'amplitude'`` uses the amplitude
         c (float): value of the phase of all displacement gates if ``execution='amplitude'``, or
             the amplitude of all displacement gates if ``execution='phase'``
+
+    Raises:
+        ValueError: if `features` or `wires` is invalid or if `method` is unknown
```
Comment (josh146) [inline pennylane/templates/embeddings.py]:
```suggestion
        ValueError: if ``features`` or ``wires`` is invalid or if ``method`` is unknown
```

File: pennylane/templates/layers.py [hunk 10]
```diff
@@ -346,6 +352,10 @@ def Interferometer(theta, phi, varphi, wires, mesh='rectangular', beamsplitter='
         beamsplitter (str): if ``clements``, the beamsplitter convention from
           Clements et al. 2016 (https://dx.doi.org/10.1364/OPTICA.3.001460) is used; if ``pennylane``, the
           beamsplitter is implemented via PennyLane's ``Beamsplitter`` operation.
+
+    Raises:
+        QuantumFunctionError: if `beamsplitter` or `mesh` is an instance of
```
Comment (josh146) [inline pennylane/templates/layers.py]:
```suggestion
        QuantumFunctionError: if ``beamsplitter`` or ``mesh`` is an instance of
```

File: pennylane/utils.py [hunk 11]
```diff
@@ -57,6 +57,9 @@ def _unflatten(flat, model):
         flat (array): 1D array of items
         model (array, Iterable, Number): model nested structure
 
+    Raises:
+        TypeError: if `model` contains an object of unsupported type
```
Comment (josh146) [inline pennylane/utils.py]:
```suggestion
        TypeError: if ``model`` contains an object of unsupported type
```

File: pennylane/utils.py [hunk 12]
```diff
@@ -79,6 +82,13 @@ def _unflatten(flat, model):
 
 def unflatten(flat, model):
     """Wrapper for :func:`_unflatten`.
+
+    Args:
+        flat (array): 1D array of items
+        model (array, Iterable, Number): model nested structure
+
+    Raises:
+        ValueError: if 'flat' has more elements than `model`
```
Comment (josh146) [inline pennylane/utils.py]:
```suggestion
        ValueError: if ``flat`` has more elements than ``model``
```

File: pennylane/qnode.py (hunk 3 above)
Comment (antalszava) [reply pennylane/qnode.py]: Added these to the current ``pennylane/qnode_new/qnode.py``!

File: pennylane/qnode.py [hunk 13]
```diff
@@ -767,6 +782,11 @@ def jacobian(self, params, which=None, *, method='B', h=1e-7, order=1, **kwargs)
             shots (int): How many times the circuit should be evaluated (or sampled) to estimate
                 the expectation values.
 
+        Raises:
+            QuantumFunctionError: if sampling is specified
+            ValueError: if `params` is incorrect, the differentiation is not possible, or if `method`
```
Comment (josh146) [inline pennylane/qnode.py]:
```suggestion
            ValueError: if `params` is incorrect, differentiation is not possible, or if `method`
```

File: pennylane/qnode.py [hunk 14]
```diff
@@ -1126,6 +1156,9 @@ def to_torch(self):
 
     def to_tf(self):
         """Convert the standard PennyLane QNode into a :func:`~.TFQNode`.
+
+        Raises:
+            QuantumFunctionError: if ``TensorFlow`` is not installed
```
Comment (josh146) [inline pennylane/qnode.py]:
```suggestion
            QuantumFunctionError: if TensorFlow >= 1.12 is not installed
```

File: pennylane/qnode.py [hunk 15]
```diff
@@ -1114,6 +1141,9 @@ def _pd_analytic_var(self, param_values, param_idx, **kwargs):
 
     def to_torch(self):
         """Convert the standard PennyLane QNode into a :func:`~.TorchQNode`.
+
+        Raises:
+            QuantumFunctionError: if ``PyTorch`` is not installed
```
Comment (josh146) [inline pennylane/qnode.py]:
```suggestion
            QuantumFunctionError: if PyTorch is not installed
```

**Pull request merged**